lib/diaspora_federation/entities/relayable_retraction.rb

```diff
@@ -46,7 +46,76 @@ class RelayableRetraction < Entity
       #   This signature is mandatory only when federation from the subscriber to an upstream
       #   author is done.
       #   @return [String] target author signature
-      property :target_author_signature
+      property :target_author_signature, default: nil
+
+      # Generates XML and updates signatures
+      # @see Entity#to_xml
+      # @return [Nokogiri::XML::Element] root element containing properties as child elements
+      def to_xml
+        entity_xml.tap do |xml|
+          hash = to_h
+          self.class.update_signatures!(hash)
+          xml.at_xpath("target_author_signature").content = hash[:target_author_signature]
+          xml.at_xpath("parent_author_signature").content = hash[:parent_author_signature]
+        end
+      end
+
+      # Adds signatures to a given hash with the keys of the author and the parent
+      # if the signatures are not in the hash yet and if the keys are available.
+      #
+      # @param [Hash] hash hash given for a signing
+      def self.update_signatures!(hash)
+        target_author = DiasporaFederation.callbacks.trigger(
+          :fetch_entity_author_id_by_guid,
+          hash[:target_type],
+          hash[:target_guid]
+        )
+        pkey = DiasporaFederation.callbacks.trigger(:fetch_private_key_by_id, hash[:diaspora_id])
+
+        fill_required_signature(target_author, pkey, hash) unless pkey.nil?
+      end
+
+      def self.fill_required_signature(target_author, pkey, hash)
+        if target_author == hash[:diaspora_id] && hash[:target_author_signature].nil?
+          hash[:target_author_signature] = Signing.sign_with_key(apply_signable_exceptions(hash), pkey)
+        elsif target_author != hash[:diaspora_id] && hash[:parent_author_signature].nil?
+          hash[:parent_author_signature] = Signing.sign_with_key(apply_signable_exceptions(hash), pkey)
+        end
+      end
+      private_class_method :fill_required_signature
+
+      # verifies the signatures (+target_author_signature+ and +parent_author_signature+ if needed)
+      # @param [Hash] data hash with data to verify
+      # @raise [SignatureVerificationFailed] if the signature is not valid or no public key is found
+      def self.verify_signatures(data)
```

Comment:
Am 02.12.15 um 02:21 schrieb Benjamin Neff:

>   * We can use |Retraction| instead of |SignedRetraction| if the current
>     diaspora can handle this. But we should also test if this is
>     compatible with friendica (cc @annando <https://github.com/annando>).

Friendica can receive retractions, signed_retractions and
relayable_retractions:
https://github.com/friendica/friendica/blob/master/include/diaspora.php#L95-L103

I haven't worked with retractions by now. Just tell me what they are
expected to do.

By now it seems as if received retractions are supported for persons and
posts. We are sending them from Friendica only for persons. for posts we
use the signed ones.

Michael